src/mlpack/methods/ann/layer/batch_norm.hpp

```diff
@@ -64,8 +64,14 @@ class BatchNorm
    *
    * @param size The number of input units / channels.
    * @param eps The epsilon added to variance to ensure numerical stability.
+   * @param average Boolean to determine whether counting average is used for
+   *                updating the parameters or momentum is used.
```

Comment:
I think they are using cumulative moving average, so we should mention that.
Also could you give me a link to the pytorch implementation that you are referring to.

Reply:
Here is the [python based](https://github.com/ptrblck/pytorch_misc/blob/master/batch_norm_manual.py) batchnorm implementation with PyTorch also the [notebook](https://colab.research.google.com/drive/12Yqa6G9fqkLukUIG5s6denFnx_gfSnr5?usp=sharing) where I use both the above implementation and PyTorch for reference. Also, Could you help in figuring why the serialization tests keep failing. Before I made the last 2 commits they were working fine and I checked that all parameters are serialized.

Reply:
Can you check which of the assertions are failing and try to see the outputs after loading the serialised model?
Does the test pass locally?

Reply:
> Does the test pass locally?

It does. According to azure build, I think it's not able to load the serialized model again (It fails on this [line](https://github.com/mlpack/mlpack/blob/ba5f9b075962740adc26260a105ce5eac27232ba/src/mlpack/tests/serialization.hpp#L235)).

Reply:
I guess we are getting `Input stream error`

Reply:
Yes, I'm trying to reproduce this locally but can't so far. I'll set up Ubuntu and see what's wrong there.

Reply:
Actually I am also on Mac OS so, won't be able to reproduce this. @KimSangYeon-DGU can you take a look at this issue?

Reply:
@saksham189 @kartikdutt18 I have an Ubuntu 18.04 machine, so I run the test and it succeeded.

Reply:
Hmm, Strange, I think I have a fix. Pushing a commit hopefully that will work.

Reply:
Oh, I deleted a previous comment because I built on `master` branch.
I'm building on `ErrorInBatchNorm` branch.

Reply:
No worries, Could you please pull from the branch again.

Reply:
Had the same issue... I need to figure out it.
```
Running 1 test case...
input stream error
/home/skim/Desktop/DEV/GSoC/Kartik/mlpack/src/mlpack/tests/serialization.hpp(235): fatal error: in "ANNLayerTest/BatchNormSerializationTest": critical check success == true has failed [false != true]

*** 1 failure is detected in the test module "mlpackTest"
```

Reply:
One think that I did figure out was that if I replace the following portion:
```
 double averageFactor = average ? 1.0 / count : momentum;
        double nElements = (double) input.n_elem / size;
        nElements = 1.0 * nElements / (1.0 * nElements - 1.0);
 
        runningMean = (1 - averageFactor) * runningMean + averageFactor * mean;
        runningVariance = (1 - averageFactor) * runningVariance + nElements *
            averageFactor * variance;
```
with the older Welford method code, It doesn't fail (The value tests will but serialization won't).

Reply:
I simplified the whole implementation and let's see the results from build (Tests pass locally). I'll add some comments in the next commit.

Reply:
I think I figured out the error, The current implementation would pass the build. The error occurs if declare the nElemenets variable which is equal to `n / (n - 1)` which is multiplied with the variance.